modules/pinger/src/main/java/org/hawkular/component/pinger/UrlChangesCollector.java

```diff
@@ -0,0 +1,157 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.hawkular.component.pinger;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+import org.hawkular.inventory.api.Action;
+import org.hawkular.inventory.api.Inventory;
+import org.hawkular.inventory.api.model.Resource;
+
+import rx.functions.Action1;
+
+/**
+ * Collects URL additions and removals reported by {@link PingManager#inventory} and synchronizes the various threads
+ * reporting the new URLs and those ones consuming them.
+ *
+ * @author <a href="https://github.com/ppalaga">Peter Palaga</a>
+ */
+public class UrlChangesCollector {
+    /**
+     * An an add or remove operation that can be performed on a {@link Set}.
+     *
+     * @author <a href="https://github.com/ppalaga">Peter Palaga</a>
+     */
+    private static class UrlChange {
+
+        /** The nature of the change - {@link Action#created()} or {@link Action#deleted()} */
+        private final Action<?, ?> action;
+
+        /** The {@link PingDestination} that the change is related to */
+        private final PingDestination destination;
+
+        /**
+         * Creates a new {@link UrlChange}.
+         *
+         * @param action the nature of the change - {@link Action#created()} or {@link Action#deleted()}
+         * @param destination the {@link PingDestination} that the change is related to
+         */
+        public UrlChange(Action<?, ?> action, PingDestination destination) {
+            super();
+            this.action = action;
+            this.destination = destination;
+        }
+
+    }
+
+    /**
+     * A callback for the {@link Inventory} that collects newly added URLs.
+     */
+    private final Action1<Resource> urlCreatedAction = new Action1<Resource>() {
+        /**
+         * It is safe to call this method concurrently from any random thread.
+         *
+         * @see rx.functions.Action1#call(java.lang.Object)
+         */
+        @Override
+        public void call(Resource r) {
+            if (PingDestination.isUrl(r)) {
+                synchronized (lock) {
+                    changes.add(new UrlChange(Action.created(), PingDestination.from(r)));
+                }
+            }
+        }
+    };
+
+    /**
+     * A callback for the {@link Inventory} that collects URL removals.
+     */
+    private final Action1<Resource> urlDeletedAction = new Action1<Resource>() {
```

Comment:
Is it possible to only use one action callback for both cases and then internally inside `call` differentiate on the incoming event?